File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int volume = 0;
        int left = 0, right = height.length - 1;
        int maxLeft = height[left], maxRight = height[right];
        
        while (left < right) {
            if (maxLeft <= maxRight) {
                left++;
                maxLeft = Math.max(maxLeft, height[left]);
                volume += maxLeft - height[left];
            } else {
                right--;
                maxRight = Math.max(maxRight, height[right]);
                volume += maxRight - height[right];
            }
        }
        
        return volume;
    }
}

File: LeetCode/Hard/0042-trapping-rain-water/0042-trapping-rain-water.kt
class Solution {
    fun trap(height: IntArray): Int {
        var volume = 0
        var left = 0
        var right = height.size - 1
        var maxLeft = height[left]
        var maxRight = height[right]

        while (left < right) {
            if (maxLeft <= maxRight) {
                left++
                maxLeft = maxOf(maxLeft, height[left])
                volume += maxLeft - height[left]
            } else {
                right--
                maxRight = maxOf(maxRight, height[right])
                volume += maxRight - height[right]
            }
        }

        return volume
    }
}